src/mme-app/ue_table.c

```diff
@@ -106,3 +113,115 @@ imsi_bin_to_str(unsigned char *b_imsi, char *s_imsi)
 	s_imsi[(BINARY_IMSI_LEN*2)-1] = '\0';
 }
 
+void init_ue_tables()
+{
+    pthread_mutex_init(&ue_link_list_mutex, NULL);
+	/* init UEs arra to 65535 initially */
+	g_UE_list[0] = (struct UE_info*)calloc(sizeof(struct UE_info), UE_POOL_CNT);
+	if(NULL == g_UE_list[0]) {
+		log_msg(LOG_ERROR, "Allocation failed.\n");
+		exit(-1);
+	}
+    ue_list_head = g_UE_list[0];
+/*
+	g_UE_list[1] = (struct UE_info*)calloc(sizeof(struct UE_info), UE_POOL_CNT);
+	if(NULL == g_UE_list[1]) {
+		log_msg(LOG_ERROR, "Allocation failed.\n");
+		exit(-1);
+	}
+	g_UE_list[2] = (struct UE_info*)calloc(sizeof(struct UE_info), UE_POOL_CNT);
+	if(NULL == g_UE_list[2]) {
+		log_msg(LOG_ERROR, "Allocation failed.\n");
+		exit(-1);
+	}
+*/
+    for(int i=0;i<10000;i++)
+        g_tmsi_allocation_array[i] = -1;
+}
+
+void add_ue_entry(struct UE_info *ue_entry)
+{
+   pthread_mutex_lock(&ue_link_list_mutex);
+   ue_entry->next_ue = ue_list_head; 
+   ue_list_head = ue_entry;
+   char imsi[16] = {'\0'}; 
+   imsi_bin_to_str(ue_entry->IMSI, imsi);
+   log_msg(LOG_INFO, "Adding IMSI = %s \n", imsi);
+   pthread_mutex_unlock(&ue_link_list_mutex);
+   return; 
+}
+
+/* Mark UE invalid, release UE index and release TMSI assigned to UE */
+void release_ue_entry(struct UE_info *ue_entry)
+{
+    struct UE_info **indirect;
+
+    /* Critical Section to manage link list */
+    pthread_mutex_lock(&ue_link_list_mutex);
+
+    indirect = &ue_list_head;
+    while((*indirect) != ue_entry)
+    {
+        indirect = &(*indirect)->next_ue;
+    }
+    *indirect = ue_entry->next_ue;
+    pthread_mutex_unlock(&ue_link_list_mutex);
+
+    ue_entry->ue_state = UNASSIGNED_ENTRY;
+    ue_entry->magic = UE_INFO_INVALID_MAGIC;
+    ue_entry->next_ue = NULL;
+
+    /* Mark tmsi free */
+    g_tmsi_allocation_array[ue_entry->m_tmsi] = -1;
+
+    /* add UE to to Pool which can be used later */
+    int ret = insert_index_into_list(ue_entry->ue_index);
+    if (ret == -1) {
+        log_msg(LOG_INFO, "List is full. More indexes cannot be added\n");
+    } else {
+        log_msg(LOG_INFO, "Index with %d is added to list\n",ue_entry->ue_index);
+    }
+}
+
+// on success return > 0 tmsi
+// on error return -1. Should never fail under fair reasonale UE load  
+// TODO : area of improvement 
+int allocate_tmsi(struct UE_info *ue_entry)
+{
+    int32_t tmsi;
+    while(1)
+    {
+      tmsi = rand() % 10000;
+      if(g_tmsi_allocation_array[tmsi] == -1)
+      {
+        g_tmsi_allocation_array[tmsi] = ue_entry->ue_index; 
+        ue_entry->ue_index = tmsi;
```

Comment:
should be : ue_entry->m_tmsi = tmsi;

Reply:
Opps..Good catch..I made a mistake here..IT should be ue_entry->m_tmsi = tmsi;